exporter/stefexporter/exporter.go

```diff
@@ -5,26 +5,395 @@ package stefexporter // import "github.com/open-telemetry/opentelemetry-collecto
+		maxAckWaitTime:      cfg.maxAckWaitTime,
+	}
+
+	exp.ackCond = sync.NewCond(&exp.ackDataMutex)
+	if exp.maxAckWaitTime == 0 {
+		exp.maxAckWaitTime = defMaxAckWaitTime
+	}
+
+	exp.compression = pkg.CompressionNone
+	if cfg.Compression == "zstd" {
+		exp.compression = pkg.CompressionZstd
+	}
+	return exp
+}
+
+func (s *stefExporter) Start(ctx context.Context, host component.Host) error {
+	s.host = host
+
+	// No need to block Start(), we will begin connection attempt in a goroutine.
+	go func() {
+		if err := s.ensureConnected(ctx, host); err != nil {
+			s.logger.Error("Error connecting to destination", zap.Error(err))
+			// exportMetrics() will try to connect again as needed.
+		}
+	}()
 	return nil
 }
 
+func (s *stefExporter) ensureConnected(ctx context.Context, host component.Host) error {
+	s.connMutex.Lock()
+	defer s.connMutex.Unlock()
+
+	if s.isConnected {
+		return nil
+	}
+
+	s.logger.Debug("Connecting to destination", zap.String("endpoint", s.cfg.Endpoint))
+
+	// Initialize ack-related fields.
+	s.lastAckedID = 0
+	s.interruptAckWaits = false
+	s.ackExpectedBy = s.ackExpectedBy[:0]
+
+	// Connect to the server.
+	var err error
+	s.grpcConn, err = s.cfg.ClientConfig.ToClientConn(ctx, host, s.set)
+	if err != nil {
+		return err
+	}
+
+	// Open a STEF/gRPC stream to the server.
+	grpcClient := stef_proto.NewSTEFDestinationClient(s.grpcConn)
+
+	// Let server know about our schema.
+	schema, err := oteltef.MetricsWireSchema()
+	if err != nil {
+		return err
+	}
+
+	settings := stefgrpc.ClientSettings{
+		Logger:       &loggerWrapper{s.logger},
+		GrpcClient:   grpcClient,
+		ClientSchema: schema,
+		Callbacks: stefgrpc.ClientCallbacks{
+			OnAck: s.onGrpcAck,
+		},
+	}
+	client := stefgrpc.NewClient(settings)
+
+	grpcWriter, opts, err := client.Connect(context.Background())
+	if err != nil {
+		s.grpcConn.Close()
+		return err
+	}
+
+	opts.Compression = s.compression
+
+	// Create record writer over gRPC stream.
+	s.stefWriter, err = oteltef.NewMetricsWriter(grpcWriter, opts)
+	if err != nil {
+		return err
+	}
+
+	// Start background timing tasks.
+	s.stopTicker = make(chan struct{})
+	go s.ticker()
+
+	s.isConnected = true
+
+	return nil
+}
+
+func (s *stefExporter) disconnect() {
+	s.connMutex.Lock()
+	defer s.connMutex.Unlock()
+
+	if !s.isConnected {
+		return
+	}
+
+	s.logger.Debug("Disconnecting...")
+
+	// Notify ticker() to stop.
+	close(s.stopTicker)
+
+	// Interrupt any pending waitForAck() call. They can't succeed anymore.
+	s.ackDataMutex.Lock()
+	s.interruptAckWaits = true
+	s.ackExpectedBy = s.ackExpectedBy[:0]
+	s.ackCond.Broadcast()
+	s.ackDataMutex.Unlock()
+
+	// If there is an existing connection close it.
+	if s.grpcConn != nil {
+		if err := s.grpcConn.Close(); err != nil {
+			s.logger.Error("failed to close grpc connection", zap.Error(err))
+		}
+	}
+
+	// Wait for all in progress wait in goroutines to get the notification that
+	// we are disconnecting and exit.
+	s.inProgressWaits.Wait()
```

Comment:
What happens if a go-routine is at line 247 when you hit this point? you will pass it, but in reality that gorutine did not yet entered the ack. So you have to ensure that happens under the lock that sends the request.

https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37564/files#diff-fb9ac6a5f4f3f6d351d846eda06763a06cc2fe8f7483964b428a4a99e45757b9R247

Reply:
This is eliminated.